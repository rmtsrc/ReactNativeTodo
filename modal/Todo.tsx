import React from 'react';

import { createContext, useState } from 'react';

type Task = { id: string; task: string; done: boolean };

type TodoList = Task[];

type TodoContext = {
  list: TodoList;
  add(task: string): void;
  remove(id: string): void;
  toggleDone(id: string): void;
};

export const TodoContext = createContext<Partial<TodoContext>>({});

const defaultTodoList = [
  {
    id: '1',
    task: 'Foo',
    done: true,
  },
  {
    id: '2',
    task: 'Bar',
    done: false,
  },
];

export function TodoProvider({ children }) {
  const [list, setList] = useState(defaultTodoList);

  return (
    <TodoContext.Provider
      value={{
        list,
        add: task =>
          setList([
            ...list,
            {
              id: `${list.length + 1}`,
              task,
              done: false,
            },
          ]),
        remove: id => setList(list.filter(item => item.id !== id)),
        toggleDone: id => {
          const index = list.findIndex(item => item.id === id);
          const item = list[index];
          setList([
            ...list.slice(0, index),
            { ...item, done: !item.done },
            ...list.slice(index + 1),
          ]);
        },
      }}
    >
      {children}
    </TodoContext.Provider>
  );
}
