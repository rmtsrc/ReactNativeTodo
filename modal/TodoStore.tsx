import React, { createContext, useReducer } from 'react';

type Task = { id: string; task: string; done: boolean };
type TodoList = Task[];
type TodoDispatch = (action: object) => void;

const defaultTodoList = [
  {
    id: '1',
    task: 'Foo',
    done: true,
  },
  {
    id: '2',
    task: 'Bar',
    done: false,
  },
];

const todoReducer = (list, action) => {
  switch (action.type) {
    case 'add':
      return [
        ...list,
        {
          id: `${list.length + 1}`,
          task: action.task,
          done: false,
        },
      ];

    case 'remove':
      return list.filter(item => item.id !== action.id);

    case 'toggleDone':
      const index = list.findIndex(item => item.id === action.id);
      const item = list[index];
      return [...list.slice(0, index), { ...item, done: !item.done }, ...list.slice(index + 1)];

    default:
      throw new Error(`Unknown action type: ${action.type}`);
  }
};

export const TodoStore = createContext<TodoList>([]);
export const TodoDispatch = createContext<TodoDispatch>(() => {});

export function TodoStoreProvider({ children }) {
  const [list, dispatch] = useReducer(todoReducer, defaultTodoList);

  return (
    <TodoDispatch.Provider value={dispatch}>
      <TodoStore.Provider value={list}>{children}</TodoStore.Provider>
    </TodoDispatch.Provider>
  );
}
