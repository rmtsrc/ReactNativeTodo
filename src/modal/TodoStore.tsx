import React, { createContext, useReducer, useEffect } from 'react';
import { AsyncStorage } from 'react-native';

type Task = { id: string; task: string; done: boolean };
type TodoList = Task[];
type TodoDispatch = (action: object) => void;

const defaultTodoList = [
  {
    id: '1',
    task: 'Example completed task',
    done: true,
  },
  {
    id: '2',
    task: 'Example todo',
    done: false,
  },
];

export const todoReducer = (list, action) => {
  let newList;
  switch (action.type) {
    case 'loaded': {
      newList = action.list === null ? [...defaultTodoList] : [...action.list];
      break;
    }
    case 'add': {
      newList = [
        ...list,
        {
          id: `${list.length + 1}`,
          task: action.task,
          done: false,
        },
      ];
      break;
    }
    case 'edit': {
      const index = list.findIndex(item => item.id === action.id);
      const item = list[index];
      newList = [...list.slice(0, index), { ...item, task: action.task }, ...list.slice(index + 1)];
      break;
    }
    case 'remove': {
      newList = list.filter(item => item.id !== action.id);
      break;
    }
    case 'toggleDone': {
      const index = list.findIndex(item => item.id === action.id);
      const item = list[index];
      newList = [...list.slice(0, index), { ...item, done: !item.done }, ...list.slice(index + 1)];
      break;
    }
    default: {
      throw new Error(`Unknown action type: ${action.type}`);
    }
  }

  async function persistData() {
    await AsyncStorage.setItem('todo-list', JSON.stringify(newList));
  }
  persistData();

  return newList;
};

export const TodoStore = createContext<TodoList>(null);
export const TodoDispatch = createContext<TodoDispatch>(() => {});

export function TodoStoreProvider({ children }) {
  const [list, dispatch] = useReducer(todoReducer, null);

  useEffect(() => {
    async function fetchData() {
      const list = await AsyncStorage.getItem('todo-list');
      dispatch({ type: 'loaded', list: JSON.parse(list) });
    }
    fetchData();
  }, []);

  return (
    <TodoDispatch.Provider value={dispatch}>
      <TodoStore.Provider value={list}>{children}</TodoStore.Provider>
    </TodoDispatch.Provider>
  );
}
