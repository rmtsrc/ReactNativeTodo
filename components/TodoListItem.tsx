import React from 'react';
import { StyleSheet } from 'react-native';
import { ListItem, CheckBox } from 'react-native-elements';
import { TodoDispatch } from '../modal/TodoStore';

export default function TodoListItem({ id, task, checked }) {
  return (
    <TodoDispatch.Consumer>
      {dispatch => (
        <ListItem
          leftElement={<CheckBox checked={checked} onPress={() => dispatch({ type: 'toggleDone', id })} />}
          title={task}
          titleStyle={checked && styles.checkedText}
          onPress={() => dispatch({ type: 'toggleDone', id })}
          rightIcon={{ name: 'delete', onPress: () => dispatch({ type: 'remove', id }) }}
          bottomDivider
          pad={0}
        />
      )}
    </TodoDispatch.Consumer>
  );
}

const styles = StyleSheet.create({
  checkedText: {
    fontStyle: 'italic',
    textDecorationLine: 'line-through',
  },
});
