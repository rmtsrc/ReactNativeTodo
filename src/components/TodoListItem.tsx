import React, { useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { ListItem, CheckBox, Input } from 'react-native-elements';
import { TodoDispatch } from '../modal/TodoStore';

export default function TodoListItem({ id, task, checked }) {
  const [editing, setEditing] = useState({ isEditing: false, text: task });

  return (
    <TodoDispatch.Consumer>
      {dispatch =>
        editing.isEditing ? (
          <View style={styles.editingInput}>
            <Input
              value={editing.text}
              onChangeText={text => setEditing({ ...editing, text })}
              onSubmitEditing={() => {
                setEditing({ ...editing, isEditing: false });
                if (editing.text !== '') {
                  dispatch({ type: 'edit', id, task: editing.text });
                }
              }}
            />
          </View>
        ) : (
          <ListItem
            leftElement={<CheckBox checked={checked} onPress={() => dispatch({ type: 'toggleDone', id })} />}
            title={task}
            titleStyle={checked && styles.checkedText}
            onPress={() => setEditing({ isEditing: true, text: task })}
            rightIcon={{ name: 'delete', onPress: () => dispatch({ type: 'remove', id }) }}
            bottomDivider
            pad={0}
          />
        )
      }
    </TodoDispatch.Consumer>
  );
}

const styles = StyleSheet.create({
  checkedText: {
    fontStyle: 'italic',
    textDecorationLine: 'line-through',
  },
  editingInput: {
    padding: 20,
    paddingLeft: 50,
  },
});
