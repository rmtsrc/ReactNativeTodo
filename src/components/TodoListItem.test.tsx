import React from 'react';
import renderer from 'react-test-renderer';
const { act } = renderer;

import TodoListItem from './TodoListItem';
import { TodoDispatch } from '../modal/TodoStore';

describe('<TodoListItem />', () => {
  const defaultProps = { id: '1', task: 'my task', checked: false };
  let wrapper = renderer.create(<TodoListItem {...defaultProps} />);
  let tree = wrapper.toJSON();

  test('renders correctly when not editing', () => {
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly when editing', () => {
    act(() => {
      const { onPress } = wrapper.toTree().rendered.props;
      onPress();
    });

    const { value } = wrapper.root.findByType('TextInput').props;
    expect(value).toBe('my task');

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('dispatches correct action when the CheckBox is pressed', () => {
    const mockDispatch = jest.fn();

    wrapper = renderer.create(
      <TodoDispatch.Provider value={mockDispatch}>
        <TodoListItem {...defaultProps} />
      </TodoDispatch.Provider>
    );

    act(() => {
      const { onPress } = wrapper.root.findAllByProps({ checked: false })[1].props;
      onPress();
    });

    expect(mockDispatch).toHaveBeenCalledWith({ type: 'toggleDone', id: '1' });
  });

  test('dispatches correct action when the delete button is pressed', () => {
    const mockDispatch = jest.fn();

    wrapper = renderer.create(
      <TodoDispatch.Provider value={mockDispatch}>
        <TodoListItem {...defaultProps} />
      </TodoDispatch.Provider>
    );

    act(() => {
      const { onPress } = wrapper.root.findByProps({ name: 'delete' }).props;
      onPress();
    });

    expect(mockDispatch).toHaveBeenCalledWith({ type: 'remove', id: '1' });
  });

  test('dispatches correct action when enter is pressed while editing', () => {
    const mockDispatch = jest.fn();

    wrapper = renderer.create(
      <TodoDispatch.Provider value={mockDispatch}>
        <TodoListItem {...defaultProps} />
      </TodoDispatch.Provider>
    );

    act(() => {
      const { onPress } = wrapper.toTree().rendered.props;
      onPress();
    });

    const { onChangeText } = wrapper.root.findByType('TextInput').props;
    act(() => {
      onChangeText('changed input value');
    });

    const { onSubmitEditing } = wrapper.root.findByType('TextInput').props;
    act(() => {
      onSubmitEditing();
    });

    expect(mockDispatch).toHaveBeenCalledWith({ type: 'edit', id: '1', task: 'changed input value' });
  });
});
