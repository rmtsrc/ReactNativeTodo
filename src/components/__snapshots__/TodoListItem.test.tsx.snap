// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoListItem /> renders correctly when editing 1`] = `
<View
  style={
    Object {
      "padding": 20,
      "paddingLeft": 50,
    }
  }
>
  <View
    style={
      Object {
        "paddingHorizontal": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        editable={true}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
        value="my task"
      />
    </View>
  </View>
</View>
`;

exports[`<TodoListItem /> renders correctly when not editing 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderBottomWidth": 0.5,
        "borderColor": "#bcbbc1",
        "flexDirection": "row",
        "padding": 14,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "margin": 5,
          "marginLeft": 10,
          "marginRight": 10,
          "opacity": 1,
          "padding": 10,
        }
      }
      testID="checkbox"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text />
      </View>
    </View>
    <View
      width={0}
    />
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "fontSize": 17,
          }
        }
        testID="listItemTitle"
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
      >
        my task
      </Text>
    </View>
    <View
      width={0}
    />
    <View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "justifyContent": "center",
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
